zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -20,21 +20,78 @@ use zebra_chain::{
 
 use DbFormatChange::*;
 
-use crate::{
-    constants::latest_version_for_adding_subtrees,
-    service::finalized_state::{DiskWriteBatch, ZebraDb},
-};
+use crate::service::finalized_state::ZebraDb;
 
 pub(crate) mod add_subtrees;
 pub(crate) mod cache_genesis_roots;
 pub(crate) mod fix_tree_key_type;
+pub(crate) mod no_migration;
+pub(crate) mod prune_trees;
+pub(crate) mod tree_keys_and_caches_upgrade;
 
 #[cfg(not(feature = "indexer"))]
 pub(crate) mod drop_tx_locs_by_spends;
 
 #[cfg(feature = "indexer")]
 pub(crate) mod track_tx_locs_by_spends;
 
+/// Defines method signature for running disk format upgrades.
+pub trait DiskFormatUpgrade {
+    /// Returns the version at which this upgrade is applied.
+    fn version(&self) -> Version;
+
+    /// Returns the description of this upgrade.
+    fn description(&self) -> &'static str;
+
+    /// Runs disk format upgrade.
+    fn run(
+        &self,
+        initial_tip_height: Height,
+        db: &ZebraDb,
+        cancel_receiver: &Receiver<CancelFormatChange>,
+    ) -> Result<(), CancelFormatChange>;
+
+    /// Check that state has been upgraded to this format correctly.
+    ///
+    /// # Panics
+    ///
+    /// If the state has not been upgraded to this format correctly.
+    fn validate(
+        &self,
+        _db: &ZebraDb,
+        _cancel_receiver: &Receiver<CancelFormatChange>,
+    ) -> Result<Result<(), String>, CancelFormatChange> {
+        Ok(Ok(()))
+    }
+
+    /// Prepare for disk format upgrade.
+    fn prepare(
+        &self,
+        _initial_tip_height: Height,
+        _upgrade_db: &ZebraDb,
+        _cancel_receiver: &Receiver<CancelFormatChange>,
+        _older_disk_version: &Version,
+    ) -> Result<(), CancelFormatChange> {
+        Ok(())
+    }
+
+    /// Returns true if the [`DiskFormatUpgrade`] needs to run a migration on existing data in the db.
+    fn needs_migration(&self) -> bool {
+        true
+    }
+}
+
+fn format_upgrades() -> Vec<Box<dyn DiskFormatUpgrade>> {
+    // Note: Disk format upgrades must be run in order.
```

Comment:
I think it would be good to check if they upgrades are really in order and panic if they're not

```suggestion
    // Note: Disk format upgrades must be run in order of database version.
```

Reply:
Applied the doc change and added a test to check that they're in order in https://github.com/ZcashFoundation/zebra/pull/9263/commits/55b36028eb45e0349f56f7ebb24088f1f6c53e4e.